src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,95 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_SIZE = 10;
+    private T[] values = (T[]) new Object[INITIAL_SIZE];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        resizingCheck();
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index != 0) {
+            indexCheck(index - 1);
+        }
+        resizingCheck();
+        T[] tempoBuffer = (T[]) new Object[size - index];
+        System.arraycopy(values, index, tempoBuffer, 0, size - index);
+        values[index] = value;
+        System.arraycopy(tempoBuffer, 0, values, index + 1, size - index);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        final T result = values[index];
+        System.arraycopy(values, index + 1, values, index, size - index - 1);
+        values[size - 1] = null;
+        size--;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((values[i] != null && values[i].equals(element))
+                    || (values[i] == null && element == null)) {
```

Comment:
```suggestion
                    || (values[i] ==  element)) {
```
якщо 2 елемента будуть null їх можна по посиланню зрівняти

Reply:
Да, да, я уже запомнил после предыдущего д/з и так больше делать не буду :)
Недосмотрел когда рефакторил.